### PR TITLE
fix bug: cannot execute squelched plan node of type: 88 (7X)

### PR DESCRIPTION
**Pull request by hyongtao-db**
Related issue: [issue 11673](https://github.com/greenplum-db/gpdb/issues/11673)

#### Bug report:

``` SQL
gpadmin=# set from_collapse_limit = 1; set join_collapse_limit = 1;;
SET
SET
gpadmin=# explain select * from pg_namespace  join lateral  (select * from aclexplode(nspacl) x join pg_authid  on x.grantee = pg_authid.oid) z on true;
                                  QUERY PLAN
-------------------------------------------------------------------------------
 Nested Loop  (cost=10000000001.20..10000000005.06 rows=81 width=288)
   ->  Seq Scan on pg_namespace  (cost=0.00..1.09 rows=9 width=117)
   ->  Hash Join  (cost=1.21..1.42 rows=9 width=171)
         Hash Cond: (x.grantee = pg_authid.oid)
         ->  Function Scan on aclexplode x  (cost=0.00..0.10 rows=10 width=41)
         ->  Hash  (cost=1.09..1.09 rows=9 width=130)
               ->  Seq Scan on pg_authid  (cost=0.00..1.09 rows=9 width=130)
 Optimizer: Postgres query optimizer
(8 rows)
 
gpadmin=# select * from pg_namespace  join lateral  (select * from aclexplode(nspacl) x join pg_authid  on x.grantee = pg_authid.oid) z on true;
ERROR:  cannot execute squelched plan node of type: 88 (execProcnode.c:606)
```

In thie SQL plan, we only select metadata from QD. In other words, there is no Motion in this plan.
As we know, squelching is maily applied to avoid deadlock with Motion nodes.
But we squelch the node even though there is no connection with any QEs. This is the cause of the bug.

#### Solutions:
We add a judgement condition 'node->state->es_sliceTable->hasMotions' in 'ExecSquelchNode(PlanState *node)'.
Almoste the same change in 6X_STABLE. Please refer to [PR 14566](https://github.com/greenplum-db/gpdb/pull/14566)

#### Extended discussion:
If there are motions in our plan, materialize operation will help us avoid the squeching action, so the above bug will be dodged.
Please refer to the result below:

``` SQL
gpadmin=# set from_collapse_limit = 1; set join_collapse_limit = 1;
SET
SET
gpadmin=#  EXPLAIN SELECT *
FROM t_wishlist AS w,
 LATERAL (SELECT *
 FROM t_product AS p join t1
 ON p.product_id = t1.c1
 WHERE p.price < w.desired_price
 ) AS x;
                                                     QUERY PLAN
--------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=10000000015.17..10033059132.46 rows=1072140000 width=95)
   ->  Nested Loop  (cost=10000000015.17..10018763932.46 rows=357380000 width=95)
         Join Filter: (p.price < (w.desired_price)::double precision)
         ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=0.00..590.67 rows=33400 width=68)
               ->  Seq Scan on t_wishlist w  (cost=0.00..145.33 rows=11133 width=68)
         ->  Materialize  (cost=15.17..972.04 rows=32100 width=27)  -- Please pay more attention here!!!!!!
               ->  Hash Join  (cost=15.17..811.54 rows=32100 width=27)
                     Hash Cond: (t1.c1 = p.product_id)
                     ->  Seq Scan on t1  (cost=0.00..355.00 rows=32100 width=4)
                     ->  Hash  (cost=11.00..11.00 rows=333 width=23)
                           ->  Redistribute Motion 3:3  (slice3; segments: 3)  (cost=0.00..11.00 rows=333 width=23)
                                 Hash Key: p.product_id
                                 ->  Seq Scan on t_product p  (cost=0.00..4.33 rows=333 width=23)
 Optimizer: Postgres query optimizer
(14 rows)

gpadmin=# SELECT *
FROM t_wishlist AS w,
 LATERAL (SELECT *
 FROM t_product AS p join t1
 ON p.product_id = t1.c1
 WHERE p.price < w.desired_price
 ) AS x;
 wishlist_id | username | desired_price | product_id |       price        |  product  | c1
-------------+----------+---------------+------------+--------------------+-----------+----
           2 | joe      |            60 |          2 | 11.080079474208375 | product 2 |  2
           2 | joe      |            60 |          3 | 22.332092171344655 | product 3 |  3
```

**Signed-off-by**: 
Yongtao Huang <yongtaoh@vmware.com> 
and Zhenghua Lyu <zlyu@vmware.com>

### CODE REVIEW

**Comment by kainwen** [comment]
Thanks a lot for your contribution!

Serveral thoughts now:

* seem like not easy to reproduce this issue without `lateral`
* Greenplum now does not support `lateral` (we have documented this)
* Squelch logic is to avoid interconnect hang (such bug is really severe in product env and super hard to debug)

Based on the above 3 points, can we fix this using a conservative way, like, if the whole plan does not need dispatch, then we do not execute squelch  at all.

**Comment by hyongtao-db** [comment]
@kainwen Thanks for your suggestion.

Totally agree with you.
It's a conservative and thorough way to avoid squelching when there is no motions in the plan.
Both main branch and 6X_STABLE branch ( [pr14566](https://github.com/greenplum-db/gpdb/pull/14566) ) are revised. Please review.

File: src/backend/executor/execAmi.c [hunk 1]
```diff
@@ -704,6 +704,10 @@ ExecSquelchNode(PlanState *node)
 	if (!node)
 		return;
 
+	// If the whole plan does not have any motions, then we do not execute squelch at all.
+	if (!(node->state->es_sliceTable) || (node->state->es_sliceTable && (!node->state->es_sliceTable->hasMotions)))
```
Comment (kainwen) [inline src/backend/executor/execAmi.c:708]: Can we test like this?

```c
if (Gp_role == GP_ROLE_DISPATCH && node->state->dispatcherState == NULL)
```

or 

```c
if (Gp_role == GP_ROLE_DISPATCH && node->state->es_interconnect_is_setup == NULL)
```

File: src/test/regress/expected/namespace_gp.out [hunk 2]
```diff
@@ -118,3 +118,95 @@ ORDER BY oid;
  7012 | pg_bitmapindex
 (4 rows)
 
+SET from_collapse_limit = 1;
+SET join_collapse_limit = 1;
+EXPLAIN ANALYSE SELECT nspname, nspacl, privilege_type, rolname FROM pg_namespace
```
Comment (kainwen) [inline src/test/regress/expected/namespace_gp.out:123]: The output is a little strange. I recommend to change a place of the test case, here not a good place.

File: src/backend/executor/execAmi.c (hunk 1 above)
Comment (hyongtao-db) [reply src/backend/executor/execAmi.c:708]: Both of these tow condition seems reasonable, but serveral test case will failed.
For example:

```
// test create_view
ERROR:  type with OID 0 does not exist
```
```
// test external table
ERROR:  external table table_master command ended with SHELL TERMINATED by signal SIGPIPE
```

File: src/test/regress/expected/namespace_gp.out (hunk 2 above)
Comment (hyongtao-db) [reply src/test/regress/expected/namespace_gp.out:123]: Agree. Is 'motion_socket.sql' OK?

**Comment by hyongtao-db** [comment]
> Thanks a lot for your contribution!
> 
> Serveral thoughts now:
> 
> * seem like not easy to reproduce this issue without `lateral`
> * Greenplum now does not support `lateral` (we have documented this)
> * Squelch logic is to avoid interconnect hang (such bug is really severe in product env and super hard to debug)
> 
> Based on the above 3 points, can we fix this using a conservative way, like, if the whole plan does not need dispatch, then we do not execute squelch at all.


After testing, I don't think it's right to skip squelching rudely if there is no any motion. These is my test result with previous patch:

**Patch**:
```
[gpadmin@localhost regress]$ git diff
diff --git a/src/backend/executor/execAmi.c b/src/backend/executor/execAmi.c
index 82315c7..39a3bd4 100644
--- a/src/backend/executor/execAmi.c
+++ b/src/backend/executor/execAmi.c
@@ -72,6 +72,7 @@
 #include "nodes/pathnodes.h"
 #include "utils/rel.h"
 #include "utils/syscache.h"
+#include "cdb/cdbvars.h"
 
 
 static bool TargetListSupportsBackwardScan(List *targetlist);
@@ -715,6 +716,10 @@ ExecSquelchNode(PlanState *node)
        if (!node)
                return;
 
+       // If the whole plan does not have any motions, then we do not execute squelch at all.
+       if (Gp_role == GP_ROLE_DISPATCH && node->state && node->state->dispatcherState == NULL)
+               return;
+               
        if (node->squelched)
                return;
```

**Test result**:
```

gpadmin=# select logseverity from gp_toolkit.__gp_log_master_ext where logseverity='LOG' limit 5;
ERROR:  external table __gp_log_master_ext command ended with SHELL TERMINATED by signal SIGPIPE (13)
DETAIL:  Command: execute:cat $GP_SEG_DATADIR/log/*.csv
gpadmin=# select logseverity from gp_toolkit.__gp_log_master_ext where logseverity='LOG';         
 logseverity
-------------
 LOG
 LOG
 LOG
 LOG

```

The only difference between these two SQLs  is the absence of `limit 5`. Node `T_LimitState` enters the function `ExecSquelchNode()` first. If so, squelching will not execute.
It meets the condition of `if (Gp_role == GP_ROLE_DISPATCH && node->state && node->state->dispatcherState == NULL)`, so return directly.
But in fact it should continue executing `ExecSquelchNode()` recursively.

So I fall back to the first commit, if there is no any motion, squelching will not execute in only ExecHashJoin(), conservatively.

**Comment by higuoxing** [comment]
I'm trying to understand the issue and the story behind it and I still need some more time on this. But it seems that 'nested loop join' and 'merge join' has similar logics as 'hash join'. Will they have similar issue?

https://github.com/greenplum-db/gpdb/blob/cbd0a9f9d25d50f847339a266cd58fb494f037ba/src/backend/executor/nodeNestloop.c#L388

https://github.com/greenplum-db/gpdb/blob/cbd0a9f9d25d50f847339a266cd58fb494f037ba/src/backend/executor/nodeMergejoin.c#L1529

**Comment by hyongtao-db** [comment]
> I'm trying to understand the issue and the story behind it and I still need some more time on this. But it seems that 'nested loop join' and 'merge join' has similar logics as 'hash join'. Will they have similar issue?
> 
> https://github.com/greenplum-db/gpdb/blob/cbd0a9f9d25d50f847339a266cd58fb494f037ba/src/backend/executor/nodeNestloop.c#L388
> 
> https://github.com/greenplum-db/gpdb/blob/cbd0a9f9d25d50f847339a266cd58fb494f037ba/src/backend/executor/nodeMergejoin.c#L1529

Agree, I think 'nested loop join' and 'merge join' will face the same problem. I will try to design a test case about it.

**Comment by zxuejing** [comment]
this issue will occur even if there is a motion in the plan.
```
set from_collapse_limit = 1; set join_collapse_limit = 1;
create table pg_namespace1 as select * from pg_namespace distributed replicated;

postgres=#  select * from pg_namespace1  join lateral  (select * from aclexplode(nspacl) x join pg_authid  on x.grantee = pg_authid.oid) z on true;
ERROR:  cannot execute squelched plan node of type: 89 (execProcnode.c:612)

postgres=# explain select * from pg_namespace1  join lateral  (select * from aclexplode(nspacl) x join pg_authid  on x.grantee = pg_authid.oid) z on true;
                                         QUERY PLAN                                 
        
------------------------------------------------------------------------------------
--------
 Nested Loop  (cost=10000000345.20..10000007845.16 rows=219600 width=275)
   ->  Gather Motion 1:1  (slice1; segments: 1)  (cost=344.00..344.00 rows=24400 wid
th=104)
         ->  Seq Scan on pg_namespace1  (cost=0.00..344.00 rows=24400 width=104)
   ->  Hash Join  (cost=1.21..1.42 rows=9 width=171)
         Hash Cond: (x.grantee = pg_authid.oid)
         ->  Function Scan on aclexplode x  (cost=0.00..0.10 rows=10 width=41)
         ->  Hash  (cost=1.09..1.09 rows=9 width=130)
               ->  Seq Scan on pg_authid  (cost=0.00..1.09 rows=9 width=130)
 Optimizer: Postgres query optimizer
(9 rows)

```

**Pull request closed without merging**